### PR TITLE
Emit better error messages for all dialogs

### LINKED ISSUE #3980: Emit better error messages for all adaptive actions
**Is your feature request related to a problem? Please describe.**
By default, the runtime host for composer is configured to echo any exceptions that are thrown back to the user as a message.  While this is helpful during the development of your bot, the error messages that are echoed back typically don't do anything to help you debug your bot.  There are a number of issues.

- Error messages don't tell you which actions generated the error.
- Simple mistakes like passing "dialog.item" versus "=dialog.item" for a property result in cryptic errors like "unexpected null encountered".
 
**Describe the solution you'd like**
All actions should wrap their internal logic with exception handlers that re-brand the thrown error to include the name and/or ID of the action raising the error.

We should add more rigorous type checking for expression properties that will give better messages telling the developer which property they miss configured.

[enhancement]

### PR DESCRIPTION
**Pull request by tomlm**
Fixes #3980 

## Description
Change dialogContext to capture exceptions and annotate exception with contextual information about where the exception happened.

## Specific Changes
* Add try/catch handlers to dialogContext methods to set Exception.Data["dialogContext'] with metadata about the context that the exception was thrown.

## Testing
* Updated existing OnError testing to test and validate for the dialog context data 

### Example exception metadata
```
{
   ...,
   "data": 
  {
       "DialogContext": 
       {
            "ActiveDialog": "SetProperty[=`foo`]",
            "Parent": "innerDialog",
           "Stack": [
               "SetProperty[=`foo`]",
               "innerDialog",
               "planningTest"
           ],
       }
}
```

### CODE REVIEW

**Comment by Stevenic** [comment]
@johnataylor should weigh in as we just had a discussion today about not creating custom exception classes.

**Comment by tomlm** [comment]
I agree, wrapping is not a good approach, I have switched to using Exception.Data to add metadata about context of exception.

---
In reply to: [647899206](https://github.com/microsoft/botbuilder-dotnet/pull/4153#issuecomment-647899206) [](ancestors = 647899206)

**Comment by cleemullins** [comment]
@tomlm  CAn you resolve the merge conflicts?

File: libraries/Microsoft.Bot.Builder.Dialogs/DialogContext.cs [hunk 1]
```diff
@@ -708,5 +678,37 @@ private async Task EndActiveDialogAsync(DialogReason reason, object result = nul
                 ObjectPath.SetPathValue(this.Context.TurnState, TurnPath.LastResult, result);
             }
         }
+
+        private void SetExceptionContextData(Exception exception)
+        {
+            if (!exception.Data.Contains(nameof(DialogContext)))
```
Comment (carlosscastro) [inline libraries/Microsoft.Bot.Builder.Dialogs/DialogContext.cs:685]: Could exception.Data be null?

File: libraries/Microsoft.Bot.Builder.Dialogs/DialogContext.cs [hunk 2]
```diff
@@ -708,5 +678,37 @@ private async Task EndActiveDialogAsync(DialogReason reason, object result = nul
...
+                var stack = new List<string>();
+                var currentDc = this;
+
+                while (currentDc != null)
+                {
+                    // (PORTERS NOTE: javascript stack is reversed with top of stack on end)
+                    foreach (var item in currentDc.Stack)
+                    {
+                        // filter out ActionScope items because they are internal bookkeeping.
+                        if (!item.Id.StartsWith("ActionScope["))
```
Comment (carlosscastro) [inline libraries/Microsoft.Bot.Builder.Dialogs/DialogContext.cs]: Consider getting this from a common constant

**Pull request merged**